lib/new_relic/agent/instrumentation/rack/chain.rb

```diff
@@ -21,8 +21,16 @@ def to_app_with_newrelic_deferred_dependency_detection
           if ::NewRelic::Agent::Instrumentation::RackHelpers.middleware_instrumentation_enabled?
             ::NewRelic::Agent.logger.info("Installing #{builder_class} middleware instrumentation")
 
-            def run_with_newrelic(app, *args)
-              run_with_tracing(app) { |wrapped_app| run_without_newrelic(wrapped_app, *args) }
+            def run_with_newrelic(app = nil, *args, &block)
+              app_or_block = app || block
+              run_with_tracing(app_or_block) do |wrapped|
+                # Rack::Builder#run for Rack v3+ supports a block, and does not
+                # support args. Whether a block or an app is provided, that
+                # callable object will be wrapped into a MiddlewareProxy
+                # instance. That proxy instance must then be passed to
+                # run_without_newrelic as the app argument.
+                block ? run_without_newrelic(wrapped, &nil) : run_without_newrelic(wrapped, *args)
```

Comment:
I don't think I've ever seen `&nil` before. Can you tell me what it means?

Reply:
Absolutely. A Ruby method will use `super` (no args specified) to pass all of its own arguments or `super()` (with args specified) to pass only the specified arguments, but in both cases the calling method's own block argument will be passed along. Because we end up with a middleware proxy object for tracing, passing it as a block won't work. So I just `nil` the original block out here and ensure that the proxy object is always passed as an argument.

The original `run` method we are tracing allows either an argument or a block (but not both) and simply treats whichever came in as the `Builder` instance's `@app` object. Given that blocks and arguments are treated identically and that the new support for a block is really just a syntax convenience thing, this approach of our always standardizing on the argument approach should work.

Beyond that, Rack v3's `run` doesn't seem to support `*args`. I'm keeping them around for the older versions, but if you've passed in a block then I know you're on v3+ and I don't use them.

Please let me know if I can rework the comment blocks to help capture any of this info better.